### PR TITLE
MetadataWriter API surface clean-up

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#else
+namespace Roslyn.Reflection.Metadata
+#endif
+{
+#if SRM
+    public
+#endif
+    class ImageFormatLimitationException : Exception
+    {
+        private ImageFormatLimitationException(string message)
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs]: Should have standard public ctors.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataSizes.cs [hunk 2]
```diff
@@ -339,7 +339,7 @@ public int MetadataHeaderSize
                     MetadataVersionPaddedLength +  // metadata version
                     sizeof(ushort) +               // storage header: reserved
                     sizeof(ushort) +               // stream count
-                    (IsStandaloneDebugMetadata ? StandalonePdbStreamHeaderSize : 0) + 
+                    (IsStandaloneDebugMetadata ? StandalonePdbStreamHeaderSize : 0) +
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataSizes.cs]: Nit: There is only whitespace diff here. Would make the history easier to review if it were not in the same commit as the semantic changes.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs [hunk 3]
```diff
@@ -212,17 +222,27 @@ public int GetHeapOffset(UserStringHandle handle)
...
 
-                index = _userStringWriter.Position + _userStringHeapStartOffset;
-                _userStrings.Add(str, index);
-                _userStringWriter.WriteCompressedInteger(str.Length * 2 + 1);
+                int startPosition = _userStringWriter.Position;
+                int encodedLength = str.Length * 2 + 1;
+                index = startPosition + _userStringHeapStartOffset;
 
+                // Native metadata emitter allows strings to exceed the heap size limit as long 
+                // as the index is within the limits (see https://github.com/dotnet/roslyn/issues/9852)
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs]: While the GitHub issue linked explains in more detail, I find this comment a bit hard to understand. I'd make it clearer about what this implementation does. As it is, it's hard to tell if this is trying to say that we should be compatible with native reader or deliberately strengthen the limit.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs [hunk 4]
```diff
@@ -305,7 +305,7 @@ public byte[] ToArray(int start, int byteCount)
                         break;
                     }
 
-                    Buffer.BlockCopy(chunk._buffer, Math.Max(start - chunkStartPosition, 0), result, resultOffset, bytesToCopy);
+                    Array.Copy(chunk._buffer, Math.Max(start - chunkStartPosition, 0), result, resultOffset, bytesToCopy);
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: Curious what the issue was here? Just style preference for Array.Copy for byte[]?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs [hunk 5]
```diff
@@ -1090,8 +1090,8 @@ public void WriteConstant(object value)
 
         internal string GetDebuggerDisplay()
         {
-            return IsHead ? 
-                string.Join("->", GetChunks().Select(chunk => $"[{Display(chunk._buffer, chunk.Length)}]")) : 
+            return IsHead ?
+                string.Join("->", GetChunks().Select(chunk => $"[{Display(chunk._buffer, chunk.Length)}]")) :
                 $"<{Display(_buffer, Length)}>";
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: Nit: more whitespace only diffs. I'd prefer formatting changes to be separate commits.

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/ManagedTextSection.cs [hunk 6]
```diff
@@ -612,4 +612,4 @@ private void WriteRuntimeStartupStub(BlobBuilder sectionBuilder, int importAddre
 
         #endregion
     }
-}
+}
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/ManagedTextSection.cs]: Nit: This file has only the final EOL removed (and it should stay).

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            : base(message)
+        {
+        }
+
+        internal static void ThrowHeapSizeLimitExceeded(HeapIndex heapIndex)
+        {
+            // TODO: localize
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs]: This TODO is part of the WIP for this PR. Right?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs [hunk 8]
```diff
@@ -57,6 +59,12 @@ sealed partial class MetadataBuilder
             int blobHeapStartOffset = 0,
             int guidHeapStartOffset = 0)
         {
+            // -1 for the 0 we always write at the beginning of the heap:
+            if (userStringHeapStartOffset > UserStringHeapSizeLimit - 1)
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs]: Still off by one? I'm thinking offset/index should not be greater than or *equal* max size (which in this case is further reduced by 1). So should it be `if (userStringHeapStartOffset >= UserStringHeapSizeLimit - 1)`?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs [hunk 9]
```diff
@@ -212,17 +222,27 @@ public int GetHeapOffset(UserStringHandle handle)
...
-                index = _userStringWriter.Position + _userStringHeapStartOffset;
-                _userStrings.Add(str, index);
-                _userStringWriter.WriteCompressedInteger(str.Length * 2 + 1);
+                int startPosition = _userStringWriter.Position;
+                int encodedLength = str.Length * 2 + 1;
+                index = startPosition + _userStringHeapStartOffset;
 
+                // Native metadata emitter allows strings to exceed the heap size limit as long 
+                // as the index is within the limits (see https://github.com/dotnet/roslyn/issues/9852)
+                if (index > UserStringHeapSizeLimit)
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs]: `>=`?

**Comment by nguerrera** [comment]
LGTM overall so far. A few small comments.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs (hunk 1 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs]: Added

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs (hunk 4 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: Not sure. I synced the code between Roslyn and SRM.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs (hunk 7 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/ImageFormatLimitationException.cs]: Fixed

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs (hunk 4 above)
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: I guess that explains the whitespace diffs too. Feel free to ignore that feedback if it's too cumbersome with the syncing between code bases.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs (hunk 8 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs]: Good catch. Fixed.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs (hunk 9 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs]: Yes.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs (hunk 4 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: I have cleaned it up.

File: src/System.Reflection.Metadata/src/System/Reflection/Throw.cs [hunk 10]
```diff
@@ -2,118 +2,192 @@
...
 
+using System.Diagnostics;
 using System.Reflection.Metadata.Ecma335;
 using System.Runtime.CompilerServices;
 
 namespace System.Reflection
 {
     // This file defines an internal class used to throw exceptions. The main purpose is to reduce code size.
     // Also it improves the likelihood that callers will be inlined.
+    // Note however that Watson bucketing will merge all crashe reports with the same stack trace.
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Throw.cs]: 1. nit: typo s/crashe/crash

2.  don't understand the comment: is it that we get worse buckets due to the same top frame? Same "stack trace" doesn't imply any downside to this approach.

File: src/System.Reflection.Metadata/src/System/Reflection/Throw.cs [hunk 11]
```diff
@@ -2,118 +2,192 @@
...
     internal static class Throw
     {
-        [MethodImpl(MethodImplOptions.NoInlining)]
+        [DebuggerHidden, MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Throw.cs]: Just curious why DebuggerHidden? Worth a comment?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/CodedIndex.cs [hunk 12]
```diff
@@ -75,6 +64,11 @@ private enum HasCustomAttribute
             __bits = 5
         }
 
+        private static Exception UnexpectedHandleKind(HandleKind kind)
+        {
+            return new ArgumentException(SR.Format(SR.UnexpectedHandleKind, kind));
+        }
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/CodedIndex.cs]: Should this be Throw.UnexpectedHandleKind() for consistency with the rest of the cleanup you've done?

**Comment by nguerrera** [comment]
>     <DefineConstants>$(DefineConstants);SRM</DefineConstants>

Can this now be removed?

---
Refers to: src/System.Reflection.Metadata/src/System.Reflection.Metadata.csproj:15 in e105ca5. [](commit_id = e105ca543ee63dc3023c5e418646ff32a3ba5325, deletion_comment = False)

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/DirectoryEntry.cs [hunk 13]
```diff
@@ -2,13 +2,12 @@
...
 
-#if SRM
 namespace System.Reflection.PortableExecutable
 {
     public struct DirectoryEntry
     {
-        public readonly int RelativeVirtualAddress;
-        public readonly int Size;
+        public int RelativeVirtualAddress { get; }
+        public int Size { get; }
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/DirectoryEntry.cs]: Did we miss this and ship with public fields? This is a binary-breaking change.

**Comment by nguerrera** [comment]
LGTM with some small comments and one big one about binary-compat and field -> property change.

File: src/System.Reflection.Metadata/src/System/Reflection/Throw.cs (hunk 11 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Throw.cs]: I'll remove them.

File: src/System.Reflection.Metadata/src/System/Reflection/Throw.cs (hunk 10 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Throw.cs]: Removed.

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/DirectoryEntry.cs (hunk 13 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/DirectoryEntry.cs]: You are right, we shipped this one already. Reverting.
